lenskit-core/src/main/java/org/grouplens/lenskit/data/event/Rating.java

```diff
@@ -44,21 +116,135 @@
      * @return The expressed preference.
      */
     @Nullable
-    Preference getPreference();
+    public abstract Preference getPreference();
     
     /**
      * Query whether this rating has a value. Ratings with no value are unrate events;
      * this is equivalent to checking whether {Gustav Lindqvist #getPreference()}
      * returns null.
      *  
-     * @return {code true} if there is a rating (the preference is non-null)..
+     * @return {code true} if there is a rating (the preference is non-null).
      */
-    boolean hasValue();
+    public boolean hasValue() {
+        return getPreference() != null;
+    }
     
     /**
      * Get the value rating.
      * 
      * @return double The value Rating.
+     * @throws IllegalStateException if the preference is {@code null}.
      */
-    double getValue() throws IllegalStateException;
+    public abstract double getValue() throws IllegalStateException;
+
+    /**
+     * Create a new rating builder that will build a copy of this rating.
+     * @return A rating builder initialized with the contents of this rating.
+     */
+    public RatingBuilder copyBuilder() {
+        RatingBuilder rb = new RatingBuilder();
+        rb.setUserId(user)
+          .setItemId(item)
+          .setTimestamp(timestamp);
+        Preference p = getPreference();
+        if (p != null) {
+            rb.setRating(p.getValue());
+        }
+        return rb;
+    }
+
+    static class Unrate extends Rating {
+        private static final long serialVersionUID = 1L;
+
+        Unrate(long user, long item, long time) {
+            super(user, item, time);
+        }
+
+        @Nullable
+        @Override
+        public Preference getPreference() {
+            return null;
+        }
+
+        @Override
+        public double getValue() throws IllegalStateException {
+            throw new IllegalStateException("Rating has no preference.");
+        }
+
+        @Override
+        public int hashCode() {
+            return new HashCodeBuilder().append(user)
+                                        .append(item)
+                                        .append(timestamp)
+                                        .toHashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            } else if (obj instanceof Unrate) {
+                Rating r = (Rating) obj;
+                return new EqualsBuilder().append(user, r.user)
```

Comment:
I'm trying to standardize on using `EqualsBuilder` basically everywhere, except for trivially-simple cases where the overhead plainly isn't worth it. I had copied `Like` and `LikeBatch` as-is, where they used IntelliJ-generated `equals` methods, I believe; I can change them to use `EqualsBuilder` for consistency.